core/src/main/java/org/apache/accumulo/core/data/Mutation.java

```diff
@@ -772,6 +772,380 @@ public void putDelete(byte[] columnFamily, byte[] columnQualifier, ColumnVisibil
+   *
+   * @since 2.0
+   */
+  public interface TimestampStep extends MutationStep {
+    MutationStep timestamp(long ts);
+  }
+
+  /**
+   * Provides methods for setting the value of a Mutation. The user can provide the value as a byte array, Value, or ByteBuffer object instance and the backend
+   * will do the necessary transformation.
+   *
+   * All MutationStep methods complete a fluent Mutation API method chain.
+   *
+   * @since 2.0
+   */
+  public interface MutationStep {
+    void set(byte[] val, boolean delete);
+
+    void set(ByteBuffer val);
+
+    void set(Value val);
+
+    void delete();
+  }
+
+  /**
+   * Returns a new FamilyStep object allowing the user to define changes to the Mutation.
+   *
+   * @return a new FamilyStep object, advancing the method chain
+   * @since 2.0
+   */
+  public FamilyStep at() {
+    return new Step();
+  }
+
+  // private inner class implementing all Step interfaces
+  private class Step implements FamilyStep {
+    byte[] columnFamily;
+    int columnFamilyLength;
+
+    byte[] columnQualifier;
+    int columnQualifierLength;
+
+    byte[] columnVisibility = null;
+    int columnVisibilityLength;
+
+    boolean hasTs = false;
+    long timestamp;
+
+    private Step() {}
+
+    // methods for changing the column family of a Mutation
+    /**
+     * Modifies the column family of a mutation.
+     *
+     * @param colFam
+     *          column family
+     * @param colFamLength
+     *          column family length
+     * @return a QualifierStep object, advancing the method chain
+     */
+    public QualifierStep family(byte[] colFam, int colFamLength) {
+      columnFamily = colFam;
+      columnFamilyLength = colFamLength;
+      return this;
+    }
+
+    /**
+     * Modifies the column family of a mutation.
+     *
+     * @param colFam
+     *          column family
+     * @return a QualifierStep object, advancing the method chain
+     */
+    public QualifierStep family(byte[] colFam) {
+      return family(colFam, colFam.length);
+    }
+
+    /**
+     * Modifies the column family of a mutation.
+     *
+     * @param colFam
+     *          column family
+     * @return a QualifierStep object, advancing the method chain
+     */
+    public QualifierStep family(Text colFam) {
+      return family(colFam.getBytes(), colFam.getLength());
+    }
+
+    /**
+     * Modifies the column family of a mutation.
+     *
+     * @param colFam
+     *          column family
+     * @return a QualifierStep object, advancing the method chain
+     */
+    public QualifierStep family(CharSequence colFam) {
+      return family(new Text(colFam.toString()));
+    }
+
+    /**
+     * Modifies the column family of a mutation.
+     *
+     * @param colFam
+     *          column family
+     * @return a QualifierStep object, advancing the method chain
+     */
+    public QualifierStep family(ByteBuffer colFam) {
```

Comment:
ByteBuffers are tricky because they have a mutable position.  It would be good to document that this method does not advance the position.  Would also be good to test this.